ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -437,33 +440,39 @@ void addToIndex(List<IndexEntry> entries, FileSpan fileSpan) throws StoreExcepti
    */
   private boolean needToRollOverIndex(IndexEntry entry) {
     Offset offset = entry.getValue().getOffset();
+    int thisValueSize = entry.getValue().getBytes().capacity();
+    int thisEntrySize = entry.getKey().sizeInBytes() + thisValueSize;
     if (validIndexSegments.size() == 0) {
       logger.info("Creating first segment with start offset {}", offset);
       return true;
     }
-
     IndexSegment lastSegment = validIndexSegments.lastEntry().getValue();
+    if (lastSegment.getVersion() != PersistentIndex.CURRENT_VERSION) {
+      logger.info("Index: {} Rolling over because the Index version has changed from {} to {}", dataDir,
+          lastSegment.getVersion(), PersistentIndex.CURRENT_VERSION);
+      return true;
+    }
     if (lastSegment.getSizeWritten() >= maxInMemoryIndexSizeInBytes) {
       logger.info("Index: {} Rolling over from {} to {} because the size written {} >= maxInMemoryIndexSizeInBytes {}",
           dataDir, lastSegment.getStartOffset(), offset, lastSegment.getSizeWritten(), maxInMemoryIndexSizeInBytes);
       return true;
     }
     if (lastSegment.getNumberOfItems() >= maxInMemoryNumElements) {
-      logger.info("Index: {} Rolling over from {} to {} because the number of items in the last segment: {} >= "
-              + "maxInMemoryNumElements {}", dataDir, lastSegment.getStartOffset(), offset, lastSegment.getNumberOfItems(),
-          maxInMemoryNumElements);
+      logger.info(
+          "Index: {} Rolling over from {} to {} because the number of items in the last segment: {} >= maxInMemoryNumElements {}",
+          dataDir, lastSegment.getStartOffset(), offset, lastSegment.getNumberOfItems(), maxInMemoryNumElements);
       return true;
     }
-    if (lastSegment.getKeySize() != entry.getKey().sizeInBytes()) {
-      logger.info("Index: {} Rolling over from {} to {} because the segment keySize: {} != entry's keysize {}", dataDir,
-          lastSegment.getStartOffset(), offset, lastSegment.getKeySize(), entry.getKey().sizeInBytes());
+    if (lastSegment.getPersistedEntrySize() < thisEntrySize) {
```

Comment:
This is just to continue to keep the store general purpose and support keys of any sizes (not just `BlobId`s) regardless of what config values are used to initialize the min bytes.